include/power/pm_policy.h

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _PM_POLICY_H_
+#define _PM_POLICY_H_
+
+#include <power/power.h>
+
+#include <string.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Define the base struct necessary for a pm_policy.
+ *
+ * Different policies can have their own struct type according with
+ * their necessity, the only requirements is the struct contains this
+ * macro in the beginning of the declaration and be statically defined
+ * using @c PM_POLICY_DEFINE.
+ */
+#define PM_POLICY_BASE \
+	struct { \
+		const char *name; \
+		const size_t len; \
+		enum power_states (*next_state_get)(struct pm_policy *policy,\
+						    int32_t ticks); \
+		bool (*low_power_devices_get)(struct pm_policy *policy, \
+					      enum power_states state); \
+	}
+
+/**
+ * @brief PM Policy definition
+ *
+ * Store information about a power management policy.
+ */
+struct pm_policy {
+	PM_POLICY_BASE;
+};
+
+/**
+ * @brief Statically define and register a power management policy
+ *
+ * The policy can be accessed outside the module where it is defined using:
+ *
+ * @code extern struct <policy_type> <policy_name>; @endcode
+ *
+ * @param policy_type The pm_policy struct used
+ * @param policy_name The name of this policy
+ * @param policy_next_state_get Callback to get the next PM state
+ * @param policy_low_power_devices_get Callback to determine whether to put
+ *        devices in low power state, given the system PM state
+ */
+#define PM_POLICY_DEFINE(policy_type, policy_name, policy_next_state_get, \
+			 policy_low_power_devices_get, ...)		\
+	Z_STRUCT_SECTION_ITERABLE_ALTERNATE(pm_policy, policy_type,	\
+					    policy_name) = {		\
+		.name = STRINGIFY(policy_name),			\
+		.len = sizeof(struct policy_type),			\
+		.next_state_get = policy_next_state_get,		\
+		.low_power_devices_get = policy_low_power_devices_get,	\
+		__VA_ARGS__						\
+	}
+
+#define PM_POLICY_FOREACH(struct_type, iterator)			\
+	extern struct struct_type _CONCAT(_##struct_type, _list_start)[]; \
+	extern struct struct_type _CONCAT(_##struct_type, _list_end)[]; \
+	for (struct struct_type *iterator =				\
+		     _CONCAT(_##struct_type, _list_start);		\
+	     ({ __ASSERT(iterator <= _CONCAT(_##struct_type, _list_end), \
+			 "unexpected list end location");		\
+		     iterator < _CONCAT(_##struct_type, _list_end); }); \
+	     iterator = (struct struct_type *)				\
+		     ((uint8_t *)iterator + iterator->len))
+
+/**
+ * @brief Function to create device PM list
+ */
+void sys_pm_create_device_list(void);
```

Comment:
hopefully this will disappear someday (list will be build time generated etc...), and it won't be necessary to mention any "PM device list" like below.

Reply:
The list of devices participating in power management cannot be static: devices may be enabled or disabled, or even (in the future) instantiated at runtime.

Reply:
disabled (or non initialized) would be silently skipped from the whole dev list. Seems simpler that way than creating a list at runtime each time you need to do something around pm.

Reply:
Agreed: having PM capture and maintain a list of devices doesn't make sense, unless the cost of skipping over devices that don't support PM becomes so excessive that the cost of maintaining a separate list is worth doing.  I'd start assuming it's not worth it.

Reply:
agreed, this is not part of the policy though. I didn't touch it yet, just move it around.